File: digits/utils/time_filters.py
# Copyright (c) 2014-2017, NVIDIA CORPORATION.  All rights reserved.
from __future__ import absolute_import

import time
from flask_babel import lazy_gettext as lgt


def print_time(t, ref_time=None):
    lt = time.localtime(t)

    # ref_time is for testing
    if ref_time is None:
        now = time.localtime()
    else:
        now = time.localtime(ref_time)

    if lt.tm_year != now.tm_year:
        return time.strftime('%b %d %Y, %I:%M:%S %p', lt).decode('utf-8')
    elif lt.tm_mon != now.tm_mon:
        return time.strftime('%b %d, %I:%M:%S %p', lt).decode('utf-8')
    elif lt.tm_mday != now.tm_mday:
        return time.strftime('%a %b %d, %I:%M:%S %p', lt).decode('utf-8')
    else:
        return time.strftime('%I:%M:%S %p', lt).decode('utf-8')


def print_time_local(t, ref_time=None):
    lt = time.localtime(t)

    # ref_time is for testing
    if ref_time is None:
        now = time.localtime()
    else:
        now = time.localtime(ref_time)

    if lt.tm_year != now.tm_year:
        return time.strftime('%Y年%m月%d日 %H:%M:%S', lt).decode('utf-8')
    elif lt.tm_mon != now.tm_mon:
        return time.strftime('%m月%d日 %H:%M:%S', lt).decode('utf-8')
    elif lt.tm_mday != now.tm_mday:
        return time.strftime('%m月%d日 %H:%M:%S', lt).decode('utf-8')
    else:
        return time.strftime('%H:%M:%S', lt).decode('utf-8')


def print_time_diff_local(diff):
    if diff is None:
        return '?'

    if diff < 0:
        return lgt('负时间')

    total_seconds = int(diff)
    days = total_seconds // (24 * 3600)
    hours = (total_seconds % (24 * 3600)) // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60

    def plural(number, name):
        return '%d %s%s' % (number, name, '' if number == 1 else '')

    def pair(number1, name1, number2, name2):
        if number2 > 0:
            return '%s, %s' % (plural(number1, name1), plural(number2, name2))
        else:
            return '%s' % plural(number1, name1)

    if days >= 1:
        return pair(days, lgt('天'), hours, lgt('小时'))
    elif hours >= 1:
        return pair(hours, lgt('小时'), minutes, lgt('分'))
    elif minutes >= 1:
        return pair(minutes, lgt('分'), seconds, lgt('秒'))
    return plural(seconds, lgt('秒'))


def print_time_diff(diff):
    if diff is None:
        return '?'

    if diff < 0:
        return lgt('Negative Time')

    total_seconds = int(diff)
    days = total_seconds // (24 * 3600)
    hours = (total_seconds % (24 * 3600)) // 3600
    minutes = (total_seconds % 3600) // 60
    seconds = total_seconds % 60

    def plural(number, name):
        return '%d %s%s' % (number, name, '' if number == 1 else 's')

    def pair(number1, name1, number2, name2):
        if number2 > 0:
            return '%s, %s' % (plural(number1, name1), plural(number2, name2))
        else:
            return '%s' % plural(number1, name1)

    if days >= 1:
        return pair(days, lgt('day'), hours, lgt('hour'))
    elif hours >= 1:
        return pair(hours, lgt('hour'), minutes, lgt('minute'))
    elif minutes >= 1:
        return pair(minutes, lgt('minute'), seconds, lgt('second'))
    return plural(seconds, lgt('second'))



def print_time_diff_nosuffixes(diff):
    if diff is None:
        return '?'

    hours, rem = divmod(diff, 3600)
    minutes, seconds = divmod(rem, 60)
    return '{:02d}:{:02d}:{:02d}'.format(int(hours), int(minutes), int(seconds))


def print_time_since(t):
    return print_time_diff(time.time() - t)

File: digits/static/js/time_filters.js
// Copyright (c) 2014-2017, NVIDIA CORPORATION.  All rights reserved.

function print_time_diff(diff) {
    if (diff < 0) {
        return 'Negative Time';
    }
    var total_seconds = Math.floor(diff);
    var days = total_seconds / (24 * 3600);
    var hours = (total_seconds % (24 * 3600)) / 3600;
    var minutes = (total_seconds % 3600) / 60;
    var seconds = total_seconds % 60;

    function plural(number, name) {
        return number + ' ' + name + (number == 1 ? '' : 's');
    }

    function pair(number1, name1, number2, name2) {
        if (number2 > 0)
            return plural(number1, name1) + ', ' + plural(number2, name2);
        else
            return plural(number1, name1);
    }

    if (days >= 1)
        return pair(Math.floor(days), gettext('day'), Math.round(hours), gettext('hour'));
    else if (hours >= 1)
        return pair(Math.floor(hours), gettext('hour'), Math.round(minutes), gettext('minute'));
    else if (minutes >= 1)
        return pair(Math.floor(minutes), gettext('minute'), Math.round(seconds), gettext('second'));
    return plural(Math.round(seconds), gettext('second'));
}

function print_time_diff_simple(diff, min_unit) {
    if (diff == 'N/A') {
        return diff;
    }
    diff = Math.max(0, diff);

    if (typeof(min_unit) === 'undefined') min_unit = gettext('second');

    var total_seconds = Math.floor(diff);
    var days = total_seconds / (24 * 3600);
    var hours = (total_seconds % (24 * 3600)) / 3600;
    var minutes = (total_seconds % 3600) / 60;
    var seconds = total_seconds % 60;

    function plural(number, name) {
        return number + ' ' + name + (number == 1 ? '' : 's');
    }

    if (days >= 1 || min_unit == 'day')
        return plural(Math.round(days), gettext('day'));
    else if (hours >= 1 || min_unit == 'hour')
        return plural(Math.round(hours), gettext('hour'));
    else if (minutes >= 1 || min_unit == 'minute')
        return plural(Math.round(minutes), gettext('minute'));
    return plural(Math.round(seconds), gettext('second'));
}

function print_time_diff_terse(diff, min_unit) {
    if (diff == 'N/A') {
        return diff;
    }
    diff = Math.max(0, diff);

    if (typeof(min_unit) === 'undefined') min_unit = gettext('second');

    var total_seconds = Math.floor(diff);
    var days = total_seconds / (24 * 3600);
    var hours = (total_seconds % (24 * 3600)) / 3600;
    var minutes = (total_seconds % 3600) / 60;
    var seconds = total_seconds % 60;

    if (days >= 1 || min_unit == 'day')
        return Math.round(days) + gettext('d');
    else if (hours >= 1 || min_unit == 'hour')
        return Math.round(hours) + gettext('h');
    else if (minutes >= 1 || min_unit == 'minute')
        return Math.round(minutes) + gettext('m');
    return Math.round(seconds) + gettext('s');
}

function print_time_diff_ago(start, min_unit) {
    if (start == 'N/A') {
        return start;
    }

    var now = Date.now() / 1000;
    var time = print_time_diff_simple(now - start, min_unit);
    return time + gettext(' ago');
}
